File: src/sparkrules/parser/ast.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any, Union

Expr = Union[
    "BinaryOp", "Not", "Identifier", "Literal", "ListExpr", "InExpr", "CallExpr", "FieldAccess"
]


class BinaryOperator(Enum):
    EQ = "=="
    NE = "!="
    LT = "<"
    LE = "<="
    GT = ">"
    GE = ">="
    AND = "and"
    OR = "or"
    IN = "in"
    NOT_IN = "not in"
    CONTAINS = "contains"
    MATCHES = "matches"


@dataclass(frozen=True, slots=True)
class Literal:
    value: Any


@dataclass(frozen=True, slots=True)
class Identifier:
    name: str

    @property
    def base(self) -> str:
        if "." in self.name:
            return self.name.split(".", 1)[0]
        return self.name


@dataclass(frozen=True, slots=True)
class ListExpr:
    items: tuple[Expr, ...]


@dataclass(frozen=True, slots=True)
class InExpr:
    left: Expr
    right: Expr
    negated: bool


@dataclass(frozen=True, slots=True)
class CallExpr:
    name: str
    args: tuple[Expr, ...]


@dataclass(frozen=True, slots=True)
class FieldAccess:
    base: str
    field: str


@dataclass(frozen=True, slots=True)
class BinaryOp:
    op: BinaryOperator
    left: Expr
    right: Expr


@dataclass(frozen=True, slots=True)
class Not:
    expr: Expr


@dataclass(frozen=True, slots=True)
class FactPattern:
    bind_name: str
    fact_type: str
    constraint: Expr | None


@dataclass(frozen=True, slots=True)
class Action:
    field_path: str
    expr: Expr


@dataclass(frozen=True, slots=True)
class RuleAst:
    name: str
    salience: int
    agenda_group: str
    activation_group: str | None
    pass_name: str | None
    group_by: tuple[str, ...]
    reason_codes: tuple[str, ...]
    stop_on_fire: bool
    when: tuple[FactPattern, ...]
    then: tuple[Action, ...]


class ParseError(ValueError):
    def __init__(self, message: str, line: int = 0, col: int = 0) -> None:
        super().__init__(message)
        self.line = line
        self.col = col

File: sparkrules_native/src/ast.rs
//! Parsed rulepack JSON (mirrors `sparkrules.parser.ast`, serializer: `sparkrules.native.ast_json`).

use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

pub const NATIVE_SCHEMA: &str = "1";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[allow(clippy::enum_variant_names)]
pub enum BinaryOperator {
    #[serde(rename = "==")]
    Eq,
    #[serde(rename = "!=")]
    Ne,
    #[serde(rename = "<")]
    Lt,
    #[serde(rename = "<=")]
    Le,
    #[serde(rename = ">")]
    Gt,
    #[serde(rename = ">=")]
    Ge,
    #[serde(rename = "and")]
    And,
    #[serde(rename = "or")]
    Or,
    #[serde(rename = "in")]
    In,
    #[serde(rename = "not in")]
    NotIn,
    #[serde(rename = "contains")]
    Contains,
    #[serde(rename = "matches")]
    Matches,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "PascalCase")]
pub enum Expr {
    Literal {
        value: JsonValue,
    },
    Identifier {
        name: String,
    },
    ListExpr {
        items: Vec<Expr>,
    },
    InExpr {
        left: Box<Expr>,
        right: Box<Expr>,
        negated: bool,
    },
    CallExpr {
        name: String,
        args: Vec<Expr>,
    },
    FieldAccess {
        base: String,
        field: String,
    },
    BinaryOp {
        op: BinaryOperator,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Not {
        expr: Box<Expr>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FactPatternJson {
    pub bind_name: String,
    pub fact_type: String,
    pub constraint: Option<Expr>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionJson {
    pub field_path: String,
    pub expr: Expr,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuleJson {
    pub name: String,
    pub salience: i32,
    pub source_order: usize,
    pub reason_codes: Vec<String>,
    pub when: Vec<FactPatternJson>,
    pub then: Vec<ActionJson>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RulePackJson {
    pub drl_hash: String,
    pub native_schema: String,
    pub rules: Vec<RuleJson>,
}

#[cfg(test)]
mod serde_smoke {
    use super::*;

    #[test]
    fn binary_op_deser() {
        let j = serde_json::json!({
            "kind": "BinaryOp",
            "op": "==",
            "left": {"kind": "Literal", "value": 1},
            "right": {"kind": "Literal", "value": 1}
        });
        let _: Expr = serde_json::from_value(j).expect("expr");
    }

    #[test]
    fn lowercase_and_op() {
        let j = serde_json::json!({
            "kind": "BinaryOp",
            "op": "and",
            "left": {"kind": "Literal", "value": true},
            "right": {"kind": "Literal", "value": false}
        });
        let _: Expr = serde_json::from_value(j).expect("and");
    }

    #[test]
    fn operators_in_contains_matches() {
        for (op, tpl) in [
            (
                "not in",
                serde_json::json!({
                    "kind": "InExpr",
                    "left": {"kind": "Literal", "value": 1},
                    "right": {"kind": "ListExpr", "items": []},
                    "negated": true
                }),
            ),
            (
                "contains",
                serde_json::json!({
                    "kind": "BinaryOp",
                    "op": "contains",
                    "left": {"kind": "Literal", "value": "hay"},
                    "right": {"kind": "Literal", "value": "a"}
                }),
            ),
            (
                "matches",
                serde_json::json!({
                    "kind": "BinaryOp",
                    "op": "matches",
                    "left": {"kind": "Literal", "value": "x"},
                    "right": {"kind": "Literal", "value": "."}
                }),
            ),
        ] {
            let _: Expr = serde_json::from_value(tpl).unwrap_or_else(|e| panic!("parse {op}: {e}"));
        }
    }
}
